### PR TITLE
Cleanup error handlers: round 4

### PR DESCRIPTION
**Pull request by nnethercote**
More `rustc_errors` cleanups. A sequel to #118933.

r? @compiler-errors

### CODE REVIEW

**Comment by nnethercote** [comment]
(Currently draft code, and includes all the commits in #119097, please ignore for now.)

**Comment by nnethercote** [comment]
Lots of commits, but they're mostly small and all self-contained.

File: compiler/rustc_errors/src/lib.rs [hunk 1]
```diff
@@ -1731,25 +1731,80 @@ impl DelayedDiagnostic {
 
 #[derive(Copy, PartialEq, Eq, Clone, Hash, Debug, Encodable, Decodable)]
 pub enum Level {
+    /// For bugs in the compiler. Manifests as an ICE (internal compiler error) panic.
+    ///
+    /// Its `EmissionGuarantee` is `BugAbort`.
     Bug,
+
+    /// This is a strange one: lets you register an error without emitting it. If compilation ends
+    /// without any other errors occurring, this will be emitted as an error. Otherwise, it will be
```
Comment (compiler-errors) [inline compiler/rustc_errors/src/lib.rs:1667]: > this will be emitted as an error

It will be emitted as a bug, I though?

Comment (nnethercote) [reply compiler/rustc_errors/src/lib.rs:1667]: We have some unclear terminology. Sometimes "error" means `Level::Error` and sometimes it means `Level::{BugDelayed,Fatal,Error,FailureNote}`, e.g. [here](https://github.com/rust-lang/rust/blob/398fd92e177cd1a47ae1bb146506953e87e73d30/compiler/rustc_errors/src/diagnostic.rs#L263-L279).

But even with that in mind, this is a clumsy sentence, I'll fix it up in a minute.

**Comment by nnethercote** [comment]
I fixed the comment.

@bors r=compiler-errors

**Comment by nnethercote** [comment]
I fixed the conflicts.

@bors r=compiler-errors

**Pull request merged**